django/db/models/fields/json.py

```diff
@@ -0,0 +1,483 @@
+                if connection.mysql_is_mariadb and isinstance(obj, str):
+                    func.append("JSON_UNQUOTE(JSON_EXTRACT(%s, '$'))")
+                else:
+                    func.append("JSON_EXTRACT(%s, '$')")
+            rhs = rhs % tuple(func)
+        return rhs, rhs_params
+
+
+class KeyTransform(Transform):
+    postgres_operator = '->'
+    postgres_nested_operator = '#>'
+
+    def __init__(self, key_name, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.key_name = key_name
+
+    def _preprocess_lhs(self, compiler, connection):
+        key_transforms = [self.key_name]
+        previous = self.lhs
+        while isinstance(previous, KeyTransform):
+            key_transforms.insert(0, previous.key_name)
+            previous = previous.lhs
+
+        lhs, params = compiler.compile(previous)
+        return lhs, params, key_transforms
+
+    def as_mysql(self, compiler, connection):
+        lhs, params, key_transforms = self._preprocess_lhs(compiler, connection)
+        json_path = self.compile_json_path(key_transforms)
+        return 'JSON_EXTRACT(%s, %%s)' % lhs, params + [json_path]
+
+    def compile_json_path(self, key_transforms):
+        path = ['$']
+        for key_transform in key_transforms:
+            try:
+                num = int(key_transform)
+                path.append('[{}]'.format(num))
+            except ValueError:  # non-integer
+                path.append('.')
+                path.append(json.dumps(key_transform))
+        return ''.join(path)
+
+    def as_oracle(self, compiler, connection):
+        lhs, params, key_transforms = self._preprocess_lhs(compiler, connection)
+        json_path = self.compile_json_path(key_transforms)
+        return "COALESCE(JSON_QUERY(%s, '%s'), JSON_VALUE(%s, '%s'))" % ((lhs, json_path) * 2), params
+
+    def as_postgresql(self, compiler, connection):
+        lhs, params, key_transforms = self._preprocess_lhs(compiler, connection)
+        if len(key_transforms) > 1:
+            return "(%s %s %%s)" % (lhs, self.postgres_nested_operator), [key_transforms] + params
+        try:
+            int(self.key_name)
+        except ValueError:
+            lookup = "'%s'" % self.key_name
+        else:
+            lookup = "%s" % self.key_name
+        return "(%s %s %s)" % (lhs, self.postgres_operator, lookup), params
+
+    def as_sqlite(self, compiler, connection):
+        return self.as_mysql(compiler, connection)
+
+
+class KeyTextTransform(KeyTransform):
+    postgres_operator = '->>'
+    postgres_nested_operator = '#>>'
+
+
+class KeyTransformTextLookupMixin:
+    """
+    Mixin for combining with a lookup expecting a text lhs from a JSONField
+    key lookup. Make use of the ->> operator instead of casting key values to
+    text and performing the lookup on the resulting representation.
+    """
+    def __init__(self, key_transform, *args, **kwargs):
+        assert isinstance(key_transform, KeyTransform)
```

Comment:
Would it be better to raise `TypeError` than to use `assert` here?

Reply:
Actually, I don't think it's needed. It was copied from `contrib.postgres`'s `JSONField`. Should I remove it, or raise a `TypeError`?

Reply:
I suppose it was put there for a reason. I'd go for `TypeError` as asserts can be optimised out.